### PR TITLE
chore(logstreams): delete logstream data after snapshot

### LINKED ISSUE #2305: Delete old data from the logstream
AT:
* the broker deletes data from the logstream before position X  if
  * there exists Y valid snapshots which include the data before X (i.e. lower processing position bound >= X). Y is the number of snapshots in the configuration-
  * all exporters confirmed a position >= X

requires #2304
requires #2306

### PR DESCRIPTION
**Pull request by miguelpires**
closes #2305

### CODE REVIEW

**Review by saig0 — CHANGES_REQUESTED** [review]
We should add a test case to verify that this works all together in the broker with some exporters.

**Comment by MiguelPires** [comment]
Thanks for the feedback @saig0. I changed the tests and also removed the callback when the exporter service stops

File: qa/integration-tests/src/test/java/io/zeebe/broker/it/DataDeleteTest.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    TestUtil.waitUntil(
+        () -> Arrays.stream(snapshotsDir.listFiles()).anyMatch(f -> !f.getName().equals("/tmp")));
+
+    // then
+    Assertions.assertThat(segmentsDir.list().length).isLessThan(segments);
+  }
+
+  private static ExporterCfg createExporter() {
+    final ExporterCfg exporterCfg = new ExporterCfg();
+    exporterCfg.setClassName(TestExporter.class.getName());
```
Comment (saig0) [inline qa/integration-tests/src/test/java/io/zeebe/broker/it/DataDeleteTest.java]: We should use a different exporter. This one is defined in a different test module.

File: qa/integration-tests/src/test/java/io/zeebe/broker/it/DataDeleteTest.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  @Rule public RuleChain ruleChain = RuleChain.outerRule(brokerRule).around(clientRule);
+  @Rule public Timeout timeout = new Timeout(60, TimeUnit.SECONDS);
+
+  @Test
+  public void shouldDeleteDataWithExporters() {
+    // given
+    final String rootPath = brokerRule.getBrokerCfg().getData().getDirectories().get(0);
+    final String snapshotDirPath = rootPath + "/partition-1/state/1_zb-stream-processor/snapshots";
+    final String segmentsDirPath = rootPath + "/partition-1/segments";
+    brokerRule.getClock().pinCurrentTime();
```
Comment (saig0) [inline qa/integration-tests/src/test/java/io/zeebe/broker/it/DataDeleteTest.java]: I see no reason why we should pin the time.

**Review by saig0 — CHANGES_REQUESTED** [review]
Currently, it can happen that the data deletion is triggered by the exporter stream processor.

```
10:36:23.224 [exporter-snapshot-director] [0.0.0.0:25602-zb-fs-workers-0] INFO  io.zeebe.logstreams.snapshot - Creation of snapshot for exporter took 48 ms.
10:36:23.229 [exporter-snapshot-director] [0.0.0.0:25602-zb-fs-workers-0] INFO  io.zeebe.logstreams - Delete data from logstream until position '4294973080' (address: '4294973336').
10:36:23.232 [exporter-snapshot-director] [0.0.0.0:25602-zb-fs-workers-0] INFO  io.zeebe.logstreams - Deleted 1 segments from log storage (0 to 1).
```

This is not ok because we need to ensure that we have enough snapshots of the workflow stream processor state :sweat_smile: 
This might be fixed when we use just one RocksDB. If there is an easy way to avoid this then we should do it. Otherwise, we need to fix it with the refactoring.

**Comment by MiguelPires** [comment]
Hey @saig0, I made another iteration. The solution to prevent data from being deleted after an exporter snapshot is not the most elegant but it should work until the refactor. I added it to issue #2381 so we don't forget to remove it

**Review by saig0 — APPROVED** [review]

**Pull request merged**